add usage examples for O & O APIs (#160)
maykinmedia/objects-api: docs/api/usage.rst

        "updateFrequency": "monthly",
        "providerOrganization": "Open data for trees",
        "documentationUrl": "http://tree-object-type.nl",
        "labels": {},
        "createdAt": "2021-03-03",
        "modifiedAt": "2021-03-03",
        "versions": [
            "http://<object-type-host>/api/v1/objecttypes/<object-type-uuid>/versions/1"
        ]
    }

You can see that ``versions`` attribute includes a list of urls to all the versions of this
object type.

Objects API
===========

Now we have an object type containing JSON schema for tree objects and we are ready to
create objects. Before going further please, make sure that you configured authorizations
in the admin:

* Objects API can access objecttypes API
* token which you use have permissions for a new object type.

Create an object
----------------

First of all let's construct tree data that match a related JSON schema in Objecttypes API:

.. code-block:: json

    {
        "boomgroep": "Solitaire boom",
        "boomhoogteactueel": 3,
        "leeftijd": 100,
        "meerstammig": false
    }

https://jsonschema.dev can be used to validate JSON data against JSON schema.

> Using the url of the created object type we can create a tree object. If we have
> GEO coordinates for our object we can also include them into the request body.
> Don't forget the required "Content-Crs" header.

Reviewer: ```suggestion
Using the URL of the created object type, we can create a tree object. If we have
geographic coordinates for our object we can also include them into the request
body. Don't forget the required "Content-Crs" header to indicate the coordinate 
system you are using.
```

.. code-block:: http

    POST /api/v1/objects HTTP/1.1
    Content-Crs: EPSG:4326

    {
        "type": "http://<object-type-host>/api/v1/objecttypes/<object-type-uuid>",
        "record": {
            "typeVersion": 1,
            "startAt": "2021-01-01",
            "data": {
                "boomgroep": "Solitaire boom",
                "boomhoogteactueel": 3,
                "leeftijd": 100,
                "meerstammig": false
            },
            "geometry": {
                "type": "Point",
                "coordinates": [4.908722727852763, 52.36991749536178]
            }
        }
    }

> An object type version is defined in ``typeVersion`` attribute. This means that we can create
> objects that match any version of the particular object type. The response contains the
> url of the object:

Reviewer: ```suggestion
The object type version is defined in ``typeVersion`` attribute. This means that we can create
objects that match any version of the particular object type but in this case we only have our 
initial version. The response contains the URL of the object:
```

.. code-block:: http

    HTTP/1.1 201 Created

    {
        "url": "http://<object-host>/api/v1/objects/<object-uuid>",
        "type": "http://<object-type-host>/api/v1/objecttypes/<object-type-uuid>",
        "record": {
            "index": 1,
            "typeVersion": 1,
            "data": {
                "leeftijd": 100,
                "boomgroep": "Solitaire boom",
                "meerstammig": false,
                "boomhoogteactueel": 3
            },
            "geometry": {
                "type": "Point",
                "coordinates": [
                    4.908722727852763,
                    52.36991749536178
                ]
            },
            "startAt": "2021-01-01",
            "endAt": null,
            "registrationAt": "2021-03-03",
            "correctionFor": null,
            "correctedBy": null
        }
    }

> When an object is being created or updated its data is always validated against
> JSON schema in the related object type. If the data doesn't match the response will
> contain 400 error.

Reviewer: ```suggestion
When an object is created or updated, its data is always validated against the
JSON schema in the related object type. If the data doesn't match, the response will
contain a HTTP 400 error.
```

For example, let's try to create the following object:

.. code-block:: http

    POST /api/v1/objects HTTP/1.1

    {
        "type": "http://<object-type-host>/api/v1/objecttypes/<object-type-uuid>",
        "record": {
            "typeVersion": 1,
            "startAt": "2021-03-03",
            "data": {
                "boomgroep": "Solitaire boom",
                "boomhoogteactueel": 2.5,
                "leeftijd": 100,
            }
        }
    }

> In the JSON schema ``boomhoogteactueel`` is integer, so the response will look like this:

Reviewer: ```suggestion
In the JSON schema ``boomhoogteactueel`` is of type integer but we provide a floating point number.
The response will look like similar to this:
```

.. code-block:: http

    HTTP/1.1 400 Bad Request

    {
        "non_field_errors": [
            "2.5 is not of type 'integer'"
        ]
    }

Retrieve an object
------------------

> Once the object is created it can always be retrieved by its url:

Reviewer: ```suggestion
Once the object is created, it can always be retrieved by its URL:
```

.. code-block:: http

    GET /api/v1/objects/<object-uuid> HTTP/1.1

    HTTP/1.1 200 OK

    {
        "url": "http://<object-host>/api/v1/objects/<object-uuid>",
        "type": "http://<object-type-host>/api/v1/objecttypes/<object-type-uuid>",
        "record": {
            "index": 1,
            "typeVersion": 1,
            "data": {
                "leeftijd": 100,
                "boomgroep": "Solitaire boom",
                "meerstammig": false,
                "boomhoogteactueel": 3
            },
            "geometry": {
                "type": "Point",
                "coordinates": [
                    4.908722727852763,
                    52.36991749536178
                ]
            },
            "startAt": "2021-01-01",
            "endAt": null,
            "registrationAt": "2021-03-03",
            "correctionFor": null,
            "correctedBy": null
        }
    }

Retrieve objects of certain object type
---------------------------------------

> Objects API supports different searches through the objects.

Reviewer: ```suggestion
The Objects API supports different filter and search options.
```

> You can filter objects on:

Reviewer: ```suggestion
You can filter objects by:
```

* object type
* data attributes (display all trees higher than 2 meters)

> * GEO coordinates (display all trees in one neighbourhood)

Reviewer: ```suggestion
* geographic coordinates or areas (display all trees in one neighbourhood)
```

> To list on a particular object type you can use ``type`` query parameter:

Reviewer: ```suggestion
To filter the list by a particular object type you can use the ``type`` query parameter:
```

.. code-block:: http

    GET /api/v1/objects?type=http://<object-type-host>/api/v1/objecttypes/<object-type-uuid> HTTP/1.1

    HTTP/1.1 200 OK

    [
        {
            "url": "http://<object-host>/api/v1/objects/<object-uuid>",
            "type": "http://<object-type-host>/api/v1/objecttypes/<object-type-uuid>",
            "record": {
                "index": 1,
                "typeVersion": 1,
                "data": {
                    "leeftijd": 100,
                    "boomgroep": "Solitaire boom",
                    "meerstammig": false,
                    "boomhoogteactueel": 3
                },
                "geometry": {
                    "type": "Point",
                    "coordinates": [
                        4.908722727852763,
                        52.36991749536178
                    ]
                },
                "startAt": "2021-01-01",
                "endAt": null,
                "registrationAt": "2021-03-03",
                "correctionFor": null,
                "correctedBy": null
            }
        }
    ]

> Retrieve a history of an object

Reviewer: ```suggestion
Retrieve the history of an object
```

-------------------------------

> Objects API supports versioning, i.e. when object is updated its previous states
> which are called here as object records can also be accessed.

Reviewer: ```suggestion
The Objects API supports versioning, i.e. when an object is updated, its previous states
can also be retrieved. In the API these are called ``records``.
```

.. code-block:: http

    GET /api/v1/objects/<object-uuid>/history HTTP/1.1

    HTTP/1.1 200 OK

    [
        {
            "index": 1,
            "typeVersion": 1,
            "data": {
                "leeftijd": 100,
                "boomgroep": "Solitaire boom",
                "meerstammig": false,
                "boomhoogteactueel": 3
            },
            "geometry": {
                "type": "Point",
                "coordinates": [
                    4.908722727852763,
                    52.36991749536178
                ]
            },
            "startAt": "2021-01-01",
            "endAt": null,
            "registrationAt": "2021-03-03",
            "correctionFor": null,
            "correctedBy": null
        }
    ]

For now we have only one record, but every time the object is changed the new record will
be created.

> Retrieve an object version for a particular date

Reviewer: ```suggestion
Retrieve an object (record) for a particular date
```

------------------------------------------------

Since there could be a difference between the real date of

> the object change and its registration in the system the Objects API support both

Reviewer: ```suggestion
the object change and its registration in the system, the Objects API support both
```

formal and material history. The formal history describes the history as it should
be (stored in the ``startAt`` and ``endAt`` attributes). The material history describes the
history as it was administratively processed (stored in the ``registeredAt``
attribute).

The query parameters ``date`` (formal history) and ``registrationDate`` (material history)

> allow for querying the RECORDS as seen from both perspectives, and can yield different results.

Reviewer: ```suggestion
allow for querying the records as seen from both perspectives, and can yield different results.
```

> For example you want to display all the object with there states actual for 2021-02-02.
> First, let's do it from formal history perspective:

Reviewer: ```suggestion
For example, if you want to display all the objects as they were on 2021-02-02, you can do this from 2 perspectives.
First, let's do it from the formal history perspective:
```

.. code-block:: http

    GET /api/v1/objects?date=2021-02-02 HTTP/1.1

    HTTP/1.1 200 OK

    [
        {
            "url": "http://<object-host>/api/v1/objects/<object-uuid>",
            "type": "http://<object-type-host>/api/v1/objecttypes/<object-type-uuid>",
            "record": {
                "index": 1,
                "typeVersion": 1,
                "data": {
                    "leeftijd": 100,
                    "boomgroep": "Solitaire boom",
                    "meerstammig": false,
                    "boomhoogteactueel": 3
                },
                "geometry": {
                    "type": "Point",
                    "coordinates": [
                        4.908722727852763,
                        52.36991749536178
                    ]
                },
                "startAt": "2021-01-01",
                "endAt": null,
                "registrationAt": "2021-03-03",
                "correctionFor": null,
                "correctedBy": null
            }
        }
    ]

> We received our tree object in the response, because formally it started existing at 2021-01-01

Reviewer: ```suggestion
We received our tree object in the response, because formally it came into existance on 2021-01-01
```

(``startAt``) and never ceased (``endAt`` is empty).

> Now let's do the same but from material history perspective:

Reviewer: ```suggestion
Now let's do the same but from a material history perspective:
```

.. code-block:: http

    GET /api/v1/objects?registrationDate=2021-02-02 HTTP/1.1

    HTTP/1.1 200 OK

    []

> Our tree object was created at 2021-03-03 (``registrationAt``), so it hasn't existed at
> 2021-02-02 yet and Object API responses with empty list.

Reviewer: ```suggestion
Our tree object was created at 2021-03-03 (``registrationAt``), so it didn't exist 
(administratively speaking) at 2021-02-02 yet. Hence, the Objects API response is an empty list.

```
